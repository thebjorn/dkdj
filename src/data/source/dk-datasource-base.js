

/*
 *  Abstract base class (interface) of the data source classes.
 */
import counter from "../../core/counter";
import Class from "../../lifecycle/coldboot/dk-class";


export class DataSource extends Class {
    constructor() {
        super();
        this.data = [];
        this.total_items = undefined;
        this.selected_items = undefined;
        this.id = this.name = counter('datasource_');
    }

    /*
     *  Fill out request with defaults for missing values.
     */
    get_defaults(request) {
        return {
            start: request.start || 0,
            end: request.end || 25,
            orphans: request.orphans || 0,
            sort: request.sort || [],
            filter: request.filter || {}
        };
    }

    /*
     *  Return a list of `dk.data.Record`s that satisfy the `request`.
     *
     *  `request` can have the following fields::
     *
     *      start:    the index (0-based) of the first requested record
     *      end:      the index of the last requested record
     *      orphans:  if there are less than this many records after end, include them too
     *      sort:     a list of {field: <fieldname>, direction: "asc"|"desc"
     *      filter:   a json dict containing filter values.
     *
     *  get_records will call callback with the requested records in the following
     *  structure:
     *
     *      recordset   ::= {"fields": field-def, "records": record-data}
     *      field-def   ::= { (fieldname_1: {"name":__, "pos":__, "title":__, "type":__, ...)* }
     *      record-data ::= { (pk_1: record_1)* }
     *      record-data ::= [ record* ]
     *      field-name  ::= [-_a-zA-Z]+
     *      field-value ::= ...
     */
    async fetch_records(request) {
        return new Promise((resolve, reject) => {
            resolve({
                fields: {},
                meta: {},
                records: {}
            });
        });
    }
    get_records(request, returns) {
        returns({
            fields: {},
            meta: {},
            records: {}
        });
    }

    get_filter_data(filter_name, returns) {
        returns({
            missing: 'data missing',
            need_to: 'need to implement get_filter_data in datasource'
        });
    }

    /*
     *  Return a list of `dk.data.Field`s representing the "columns" of the
     *  data source (in the correct order).
     */
    get_fields() { return []; }

    /*
     *  Update the data source with `changes`.
     *
     *  `changes` is a list where items have the following fields::
     *
     *      {pk: <primary-key>, (field-name: new-value)+}
     *
     *   the data source should unconditionally save changes to all the specified
     *   fields, but only to the records/fields that have changes.
     */
    update(changes) {}

    /*
     *  (not implemented yet)
     *  The data-grid, via `dk.data.Data`, calls this method to warn the data
     *  source that it is about to start editing the data with primary key(s?)
     *  `pk` (so e.g. a select for update or similar can be performed).
     */
    start_editing(pk) {}
    stop_editing(pk) {}
}
