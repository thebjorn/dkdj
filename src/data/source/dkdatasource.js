//
//
// module.exports = {
//     Source: require('./dk-datasource-base.js'),
//     ArraySource: require('./dk-array-datasource.js'),
//     JSONDataSource: require('./dk-json-datasource.js'),
//     AjaxDataSource: require('./dk-ajax-datasource.js')
// };
