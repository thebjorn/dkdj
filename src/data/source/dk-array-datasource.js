
/*
 *  Use an array of pojos as a data source.
 *
 *  Usage::
 *
 *      const ds = dk.data.ArraySource.create({
 *          data: [
 *              {project: 'Generelt NT', work: '1:03:57'},
 *              {project: 'Tiktok',      work: '2:44:57'},
 *              {project: 'AFR-support', work: '1:06:43'}
 *          ]
 *      });
 */
import {DataSource} from "./dk-datasource-base";
import {sort_by} from "../utils";


export class ArraySource extends DataSource {
    constructor(arr) {
        super();

        if (Array.isArray(arr)) {
            // ArraySource([ {..}, {..}, ..])
            this.data = arr;
        } else {
            // ArraySource({ data: [..]})
            this.data = arr.data;
        }
        // data source items need to have a key
        this.data.map(function (item, i) {
            item.pk = item.pk || i;
        });
    }

    /*
     *  Update the data source with `changes`.
     *
     *  `changes` is a dict with the following shape::
     *
     *      {
     *          <primary-key>: {
     *              <field-name>: {oldval: a, newval: b}
     *              <field-2>: ...
     *          }, ...
     *      }
     *
     *   the data source should unconditionally save changes to all the specified
     *   fields, but only to the records/fields that have changes.
     */
    update(changes) {
        const self = this;
        Object.keys(changes).forEach(function (pk) {
            const change = changes[pk];
            const datarec = self.data[pk];
            for (const attr in change) if (attr !== 'pk' && change.hasOwnProperty(attr)) {
                datarec[attr] = change[attr].newval;
            }
        });
    }

    /*
     *  Return a list of `dk.data.Record`s that satisfy the `request`.
     *
     *  `request` can have the following fields::
     *
     *      start:    the index (0-based) of the first requested record
     *      end:      the index of the last requested record
     *      orphans:  if there are less than this many records after end, include them too
     *      sort:     a list of {field: <fieldname>, direction: "asc"|"desc"
     *      filter:   a json dict containing filter values.
     *
     *  get_records will call callback with the requested records in the following
     *  structure:
     *
     *      recordset   ::= {"fields": field-def, "records": record-data}
     *      field-def   ::= { (fieldname_1: {"name":__, "title":__, "type":__, ...)* }
     *      record-data ::= [ record* ]
     *      record      ::= { (field-name: field-value}* }
     *      field-name  ::= [-_a-zA-Z]+
     *      field-value ::= ...
     *
     *  or structurally:
     *
     *      recordset ::= {
     *          fields: {
     *              fieldname1: {
     *                  name: ___,
     *                  label: ___,
     *                  type: ___,
     *                  ...
     *              }, ...
     *          },
     *          meta: {
     *              start_recnum: __,
     *              totcount: __            // total number of records
     *          },
     *          records: [
     *              {field_name: field_value, ...},
     *              ...
     *          ]
     *      }
     */
    async fetch_records(request) {
        const p = request;
        p.end = (this.data.length - p.orphans < p.end) ? this.data.length : p.end;
        p.start = (p.start > p.end) ? p.end : p.start;

        this.do_sort(p);
        const search_recs = this.do_search(request);
        const result_recs = search_recs.slice(p.start, p.end);

        return new Promise((resolve, reject) => {
            resolve({
                fields: this.get_fields(),
                meta: {
                    totcount: this.data.length,
                    filter_count: search_recs.length,
                    start_recnum: p.start,
                    end_recnum: p.start + result_recs.length - 1
                },
                records: result_recs
            });
        });        
    }
    get_records(request, returns) {
        const p = request;
        p.end = (this.data.length - p.orphans < p.end) ? this.data.length : p.end;
        p.start = (p.start > p.end) ? p.end : p.start;

        this.do_sort(p);
        const search_recs = this.do_search(request);
        const result_recs = search_recs.slice(p.start, p.end);

        returns({
            fields: this.get_fields(),
            meta: {
                totcount: this.data.length,
                filter_count: search_recs.length,
                start_recnum: p.start,
                end_recnum: p.start + result_recs.length - 1
            },
            records: result_recs
        });
    }

    do_search(request) {
        if (!request.search) return this.data;
        return this.data.filter(function (rec) {
            for (let key in rec) if (rec.hasOwnProperty(key)) {
                if (typeof rec[key] === 'string' && rec[key].contains(request.search)) return true;
            }
            return false;
        });
    }

    do_sort(request) {
        if (request.sort && request.sort.length > 0 && request.sort[0].field) {
            this.data = sort_by(this.data, function (rec) {
                return rec[request.sort[0].field];
            });
            if (request.sort[0].direction === 'desc') {
                this.data.reverse();
            }
        }
    }

    /*
     *  Return a list of `dk.data.Field`s representing the "columns" of the
     *  data source (in the correct order).
     */
    get_fields() {
        if (this.data.length > 1) {
            const first = this.data[0];
            const field_defs = {};
            Object.keys(first).forEach(function (key, i) {
                field_defs[key] = {
                    name: key,
                    pos: i,
                    type: typeof first[key]
                };
            });
            return field_defs;
        }
        return {};
    }
}
