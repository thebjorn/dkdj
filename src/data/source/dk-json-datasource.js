
import {DataSource} from "./dk-datasource-base";
import {sort_by} from "../utils";


export class JSONDataSource extends DataSource {
    constructor(props) {
        super();
        props = Object.assign({}, props);
        this.data = props.data || {};
        this._column_number = {};
    }
    _keymap() {
        // returns a mapping from primary key to row
        const res = {};
        this.data.rows.forEach(function (row) {
            res[row.k] = row;
        });
        return res;
    }

    _field_order() {
        const res = {};
        this.data.cols.forEach(function (col, i) {
            res[col.name] = i;
        });
        return res;
    }
    update(changes) {
        const keymap = this._keymap();
        const fieldmap = this._field_order();

        Object.keys(changes).forEach(function (pk) {
            const change = changes[pk];
            if (change.oldval === change.newval) return;

            const datarec = keymap[pk];
            Object.keys(change).forEach(function (fieldname) {
                datarec.c[fieldmap[fieldname]] = change.newval;
            });
        });
    }

    do_sort(request) {
        if (request.sort.length && request.sort[0].field) {
            if (request.sort[0].field) {
                this.data.rows = sort_by(this.data.rows, rec => {
                    const colval = rec.c[this._column_number[request.sort[0].field]];
                    return colval.v || colval.f || colval;
                });
                if (request.sort[0].direction === 'desc') {
                    this.data.rows.reverse();
                }
            }
        }
    }

    rows_to_records(request, rows, fields) {
        return rows.map(function (row, i) {
            const record = {
                pk: row.k,
                rownum: request.start + i
            };
            Object.keys(fields).forEach(function (fname, i) {
                record[fname] = row.c[i];
            });
            return record;
        });
    }

    /*
     *  Convert json data into a recordset for use with the widgets.
     *
     *  The meta property has the following components:
     *
     *    totcount:         the total number of records in the domain
     *    filter_count:     the number of records left after applying filters.
     *                      If filter_count == totcount then no filters have
     *                      been applied.
     *    start_recnum:     the first index of the returned records
     *    end_recnum:       the last index of the returned records
     *
     */
    _get_records(request, data) {
        const fields = this.get_fields(data);
        return {
            fields: fields,
            meta: data.info,
            records: this.rows_to_records(request, data.rows, fields)
        };
    }
    
    async fetch_records(request) {
        const p = this.get_defaults(request);
        this.data.info = {
            start_recnum: p.start,
            totcount: this.data.rows.length
        };
        this.do_sort(p);
        // filter data here..
        this.data.info.filter_count = this.data.rows.length;
        // slice data here..
        this.data.rows = this.data.rows.slice(p.start, p.end);
        return new Promise((resolve, reject) => {
            try {
                const res = this._get_records(request, this.data);
                resolve(res);
            } catch (e) {
                reject(e);
            }
        });
    }

    get_records(request, returns) {
        const p = this.get_defaults(request);
        this.data.info = {
            start_recnum: p.start,
            totcount: this.data.rows.length
        };
        this.do_sort(p);
        // filter data here..
        this.data.info.filter_count = this.data.rows.length;
        // slice data here..
        this.data.rows = this.data.rows.slice(p.start, p.end);
        returns(this._get_records(request, this.data));
    }
    get_fields(data) {
        // returned as part of _get_records(..)
        const self = this;
        const field_defs = {};
        if (!data || !data.cols) return [];
        data.cols.forEach(function (col, i) {
            self._column_number[col.name] = i;
            field_defs[col.name] = {
                name: col.name,
                pos: i,
                type: col.datatype,
                sortable: col.sortable,
                label: col.label,
                help_text: col.help_text,
                widget: col.widget,
                data: col.data
            };
        });
        return field_defs;
    }
}
