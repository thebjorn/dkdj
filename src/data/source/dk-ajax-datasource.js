

import {JSONDataSource} from "./dk-json-datasource";
import is from "../../is";
import {dkconsole} from "../../lifecycle/dkboot/dk-console";
import jason from "../datacore/dk-json";
import {ajax, async_ajax} from "../../browser/dk-client";
import {DataQuery} from "../dk-dataquery";


export class AjaxDataSource extends JSONDataSource {
    constructor(props) {
        super();
        props = Object.assign({}, props || {});
        this.url = props.url || "";
        this._fields = null;
        this.default_pagesize = props.default_pagesize || 5;
    }

    _axdata(request) {
        if (request == null) request = new DataQuery();
        const sortcol = function (sitem) {
            if (sitem.field === undefined) return '';
            return (sitem.direction === 'desc' ? '-' : '') + sitem.field;
        };
        
        const data = {
            s: request.sort.map(sortcol).join(','),
            start: request.start,
            q: request.search,
            end: request.end,
            ft: request.filter
        };
        if (data.s === "") delete data.s;
        if (data.start === 0) delete data.start;
        if (data.q === "") delete data.q;
        if (is.isEqual(data.ft, {})) delete data.ft;
        return data;
    }
    
    _make_ajax_opts(cmd, data) {
        // for data to be sent as json, the type must be POST, dataType 'json', contentType
        // set as below, and data must be json (ie. a string).
        const self = this;
        const defaults = {
            cache: false,
            type: 'POST',
            dataType: 'json',
            url: self.url + '!' + cmd,
            contentType: "application/json; charset=utf-8",
            statusCode: {
                404() { dkconsole.debug("Page not found: " + self.url); },
                500() { window.open(self.url); }
            },
            error(req, status, err) {
                self.waiting = false;
                dkconsole.warn("ERROR", req, status, err);
                //dk.notify(self, 'fetch-data-error', req, status, err);
                throw {error: "fetch-error", message: status + ' ' + err};
            },
            converters: {
                "text json": jason.parse
            },
            success(data) {
                // dk.debug("dk-ajax-datasource-DATA:", data);
                if (data.error) {
                    dkconsole.warn(data);
                    throw data;
                } else {
                    //self.data = json.postparse(data);
                    self.data = data;
                    self.waiting = false;
                    this.do_success(self.data);
                }
            }
        };
        Object.assign(defaults, data);
        defaults.data = JSON.stringify(defaults.data);
        return defaults;
    }

    _ajax(cmd, data) {
        ajax(this._make_ajax_opts(cmd, data));
    }

    update(changes, fn) {
        // const keymap = this._keymap();
        // const fieldmap = this._field_order();
        const self = this;

        this._ajax('update', {
            data: {
                data: this.data,
                changes: changes
            },
            do_success(data) {
                fn(self._get_records(self.last_request, data), self.last_request);
            }
            // do_success(changes) {
            //     console.log("update-data", data);
            //     Object.keys(changes).forEach(function (pk) {
            //     const change = changes[pk];
            //     const datarec = keymap[pk];
            //     Object.keys(change).forEach(function (fieldname) {
            //         datarec.c[fieldmap[fieldname]] = change.newval;
            //     });
            // });
            // }
        });
    }

    get_filter_data(name, callback) {
        this._ajax('get-filter-values', {
            data: {name: name},
            do_success(data) {
                callback(data);
            }
        });
    }

    async fetch_records(request) {
        const self = this;
        //const p = this.get_defaults(request);
        this.last_request = request;
        const axdata = self._axdata(request);
        const data = await async_ajax(this._make_ajax_opts('get-records', {
            data: axdata,
        }));
        return this._get_records(request, data);
    }
    
    get_records(request, returns) {
        const self = this;
        //const p = this.get_defaults(request);
        this.last_request = request;
        const axdata = self._axdata(request);

        this._ajax('get-records', {
            data: axdata,
            do_success(data) {
                returns(self._get_records(request, data));
            }
        });
    }
}
