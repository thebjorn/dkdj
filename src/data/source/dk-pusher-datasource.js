import {AjaxDataSource} from "./dk-ajax-datasource";
// import $ from 'jquery';
import Pusher from 'pusher-js';
import {ArraySource} from "./dk-array-datasource";
import dk from '../../dk-obj'

// This datasource support fetching data from Pusher.
export class PusherDataSource extends AjaxDataSource {
// export class PusherDataSource extends ArraySource {
    constructor(props) {
        // super([
        //     {username: 'hello', age: 42},
        //     {username: 'hello2', age: 43},
        // ]);
        super();
        props = Object.assign({}, props || {});
        this.datasource_url = props.url;
        this.pusher_channel = props.pusher_channel;
        this.pusher_key = props.pusher_key;
        this._fields = null;
        this.default_pagesize = props.default_pagesize || 5;
        const pusher = new Pusher(this.pusher_key, {
            cluster: 'eu'
        });
        this.channel = pusher.subscribe(this.pusher_channel);
        
        /*
            The update field message has the following format:
    
                 {
                    'id': 233332
                    'fields': [
                        {
                            'fieldname': 'status',
                            'value': 'logged-in'
                        }*
                    ]
                }
         */
        this.channel.bind('change', data  => {
            dk.trigger(this, 'update-record', data.id, data.fields)
        });
        // this.channel.bind('new', data  => this.trigger('reload-all', data));
        // this.channel.bind('delete', data  => this.trigger('delete-record', data));
        // this.channel.bind('reload', data  => this.trigger('reload-all', data));
        // this.channel.bind('reload_record', data  => this.trigger('update-record', data));
    }
    set live_data(val) {
        this.channel.bind('new', data  => this.on_new_event(data));
        // this.channel.bind('change', data  => this.on_change_event(data));
        this.channel.bind('delete', data  => this.on_delete_event(data));
        this.channel.bind('reload', data  => this.reload_table_data());
        this.channel.bind('reload_record', data  => this.on_reload_record(data));
    }
    
    reload_table_data() {
        // this.caller is set in ResultSet.construct
        this.caller.table.table_data.pages = {};
        this.caller.table.table_data.get_page();
    }
    
    // change_fields(data) {
    //     // For now, we only update one field.
    //     // to-do: Loop through changed elements.
    //     // to-do: Surely there is an easier way of doing this?
    //     const record = this.caller.table.table_data.get_record(data['id']);
    //     const row = this.caller.table.rows[record.rownum];
    //     const _row = $('#' + row.id);
    //     data['fields'].forEach(field => {
    //         const column = this.caller.table.column[field['fieldname']];
    //         const cell = $("#" + row.cells[column.colnum].id);
    //         cell.text(field['value']);
    //     });
    //    
    //     // animate the record change
    //     // fade-in
    //     _row.addClass('record_change');         
    //     setTimeout(function() {
    //         // fade-out
    //         _row.addClass('record_change_done');
    //     }, 1000);
    //    
    //     setTimeout(function() {
    //         _row.removeClass('record_change');
    //         _row.removeClass('record_change_done');
    //     }, 2000);
    // }
    
    // Change these functions if we want to do something else than just to 
    // reload the datasource.
    
    // One thing we should implement is to have different animation-effects
    // for the different events. 
    // ref: https://pusher.com/tutorials/realtime-table-datatables/#setting-up-datatables
    
    // PS! Reloading the datasource executes generates db-hits, so updating the 
    // fields directly using the incoming data, is the way to go except when
    // reloading using a scheduled job (because of non-trigger updates ie. 
    // time-based updates).
    
    on_new_event(data) {
        this.reload_table_data();
    }
    on_change_event(data) {
        this.change_fields(data);
    }
    on_delete_event(data) {
        this.reload_table_data();
    }
    on_reload_record(data) {
        
    }
}
