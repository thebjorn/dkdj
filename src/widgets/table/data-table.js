/*
 *  A DataTable is the normal table representation of a DataSource,
 *  i.e. records in rows, with each field in its own column (with a column
 *  header).
 */

import dk from "../../dk-obj";
import {TableLayout} from "../../layout/table-layout";
import {Widget} from "../../widgetcore/dk-widget";
import {TableRow} from "./table-row";
import is from "../../is";
import {ColumnDef} from "./column-def";
import {TableHeader} from "./table-header";
import {DataSet} from "../../data/dk-dataset";
import {AjaxDataSource} from "../../data/source/dk-ajax-datasource";
import {dkwarning} from "../../lifecycle/coldboot/dkwarning";


export class DataTableLayout extends TableLayout {
    constructor(widget, location, template, structure) {
        super(widget, location, template, structure);
        this.thead_tr = this.add_header_row();
    }
    
    add_thead_th(attrs) {
        const th = this.make('thead_th', attrs);
        this.thead_tr.appendln(th);
        return th;
    }
}


export class DataTable extends Widget {
    constructor(...args) {
        super({
            // type: 'datatable',
            dklayout: DataTableLayout,
            TableRow: TableRow,
            current_pagenum: 0,
            translate: {},  // l10n

            structure: {
                thead: {},
                tfoot: {},
                tbody: {}
            },

            template: {
                root: 'table',
                table: '<table/>',
                thead: '<thead/>',
                thead_tr: '<tr/>',
                thead_th: '<th/>',
                tbody: '<tbody/>',
                tbody_tr: '<tr/>'
            },

            table_data: null,
            state: undefined,

            // The columns attribute should be a property object that defines the visible columns:
            //
            //      columns: {
            //          colnam1: {}
            //          colname2: {..}
            //      ..
            //
            // this.columns is converted to an array of dk.table.ColumnDef, and this.column to a
            // dict from column name to dk.table.ColumnDef.
            columns: null,
            column: {},

            filters: {},

            // should table be sortable?
            sortable: true,

            // selector of dom element to bless as a download link (e.g.
            // <a id="foo" data-download="filename.csv">download</a>
            // =>  download: '#foo',
            download: undefined,

            header: null,
            rows: null,

            column_order: null,
            column_defs: null,            
        }, ...args);

        const self = this;
        this.rows = [];
        this.header = [];   // th cell objects

        this.column = {};
        this.column_order = [];
        if (!!this.datasource && this.table_data === null) {
            const dsprops = {datasource: this.datasource};
            if (this.pagesize != null) dsprops.pagesize = this.pagesize;
            if (this.orphans != null) dsprops.orphans = this.orphans;
            if (this.pagenum != null) dsprops.pagenum = this.pagenum;
            this.table_data = new DataSet(dsprops);
        }
        if (this.table_data === null) {
            this.data_url = (this.dk && this.dk.url)? this.dk.url : "";
            //dk.info("URL", this.data_url);
            // if data is not defined, then create one
            this.table_data = new DataSet({
                // default data source is AjaxDataSource
                datasource: new AjaxDataSource({
                    // default url is current url
                    url: self.data_url
                })
            });
        }
        if (!this.data_url) this.data_url = this.table_data.datasource.url;

        //        this.state = dk.State.create({
        //            state: this.table_data.current_pagedef,
        //            engine: dk.HashStorage.create({})
        //        });
        //this.state.restore();
    }

    construct() {
        // this.state = dk.page.hash.substate(this.id);

        if (this.download) {
            const $download = dk.$(this.download);
            const filename = $download.attr('data-download') || 'filename.csv';
            dk.$(this.download).attr('href', this.data_url + "!get-records?fmt=csv&filename=" + filename);
        }
    }

    /*
     *  Connect to a pager widget.
     */
    set_pager(pwidget) {
        // const self = this;
        if (this.table_data.page) {
            pwidget.set_pagecount(this.table_data.page.recordset.meta.pagecount);
            pwidget.select_page(this.table_data.page.query.pagenum);
        }

        dk.on(pwidget, 'select-page', (...args) => this.table_data.get_pagenum(...args));
        dk.on(this.table_data, 'fetch-info', (info, query) => {
            pwidget.set_pagecount(info.pagecount);
            pwidget.select_page(query.pagenum);
        });
    }

    filter_values() {
        return {};
    }

    handlers() {
        //$(this.table_data).on('fetch-data', _.bind(this.draw, this));
        //$bind('fetch-data@data -> draw@me', {data: this.table_data, me: this});
        dk.on(this.table_data, 'fetch-data-start', () => this.start_busy());
        dk.on(this.table_data, 'fetch-data', (...args) => this.draw(...args));
        dk.on(this.table_data, 'fetch-data', (...args) => this.end_busy(...args));
    }

    set_sort(sortcol) {
        if (!sortcol.sortable) return;
        if (this.sortable) {
            this.header.forEach(function (col) {
                if (col.sortable && col !== sortcol) col.sort_icon.clear_sort();
            });
            sortcol.sort_icon.toggle_sort_direction();
            this.table_data.set_sort(sortcol.name, sortcol.sort_icon.direction);
            dk.trigger(this, 'sort-order', sortcol.name, sortcol.sort_icon.direction, this);
        }
    }

    set_search(terms) {
        this.table_data.set_search(terms);
        if (this.state) this.state.set('pagedef', this.table_data.current_pagedef);
    }

    set_filter(vals) {
        this.table_data.set_filter(vals);
    }

    get_xy(colnum, rownum) {
        return this.rows[rownum].cells[colnum];
    }

    set_xy(colnum, rownum, val, html, title) {
        const cell = this.get_xy(colnum, rownum);
        cell.set_value(val, html, title);
        return cell;
    }

    values() {
        const res = [];
        this.rows.forEach(function (row) {
            const r = [];
            row.cells.forEach(function (cell) {
                r.push(cell.widget().text());
            });
            res.push(r);
        });
        return res;
    }

    /*
     *  Convert this.columns to an array of basic property objects.
     *
     *    {foo: {label: 'Foo', ..}} ==> [{name: 'foo', label: 'Foo', ...}, ...]
     */
    _get_columndefs(fieldlst) {
        const self = this;
        let columndefs;
        const fields = {};  // field.name -> field
        fieldlst.forEach(function (f) { fields[f.name] = f; });

        // make sure this.columns is an array of pojos
        if (is.isFunction(this.columns)) columndefs = this.columns(this.table_data.fields);
        else if (this.columns === null) columndefs = fieldlst;
        else if (!Array.isArray(this.columns)) {
            // the columns property has been defined, convert it to array
            columndefs = Object.keys(this.columns).map(name => {
                const col = this.columns[name];
                col.name = name;
                return col;
            });
        } else {
            dkwarning(`Expected {field1: {...}, field2: {...}, etc.}. columns property of unknown type: ${this.columns}`);
        }

        // convert the pojos to ColumnDef objects
        return columndefs.map(function (coldef, colnum) {
            const field = fields[coldef.name];
            const tmp = {};
            if (field) {
                Object.assign(tmp, {
                    fieldpos: field.pos,
                    fieldtype: field.type,
                    sortable: field.sortable,
                    label: field.label,
                    help_text: field.help_text,
                    field_widget: field.widget,
                    field_data: field.data
                });
            }
            Object.assign(tmp, coldef);
            Object.assign(tmp, {
                colnum: colnum,
                table: self
            });
            if (tmp.sortable === undefined) {
                tmp.sortable = self.sortable;
            }
            const cdef = new ColumnDef(tmp);
            if (fields[cdef.name]) cdef.bind_to_field(fields[cdef.name]);
            self.column[cdef.name] = cdef;
            self.column_order.push(cdef.name);
            return cdef;
        });
    }

    /*
     *  Draw the column headers.
     */
    draw_header(dataset) {  // must be a separate method so VDataTable can inherit..
        if (!this.drawn_header) {
            const self = this;
            // NOTE: we have data here!
            this.columns = this._get_columndefs(dataset.page.fields);

            this.columns.forEach(function (coldef) {
                const th = TableHeader.create(self.layout.add_thead_th(), coldef);
                self.header.push(th);
            });
            self.drawn_header = true;  // only draw header once.
        }
    }

    draw(dataset) {
        const self = this;

        if (!dataset) {   // NOTE: data source consumer pattern..?  (on: data -> draw?)
            this.table_data.get_page({
                pagenum: this.current_pagenum
            });
        } else {
            self.draw_header(dataset);
            dk.trigger(self, 'draw-start', self);
            self.layout.clear_body();
            self.rows = [];

            dataset.page.records.forEach(function (record, rownum) {
                const tr = self.TableRow.create(self.layout.add_row_to('tbody'), {
                    rownum: rownum,
                    record: record,
                    table: self
                });
                self.rows.push(tr);
            });
            dk.trigger(self, 'draw-end', self);
        }
    }
}
