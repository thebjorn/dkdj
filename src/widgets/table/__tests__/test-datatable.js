
import $ from 'jquery';
import utidy from "../../../browser/dk-html";
import page from "../../../widgetcore/dk-page";
import {ArraySource} from "../../../data/source/dk-array-datasource";
import {DataTable} from "../data-table";


test("test-datatable", () => {
    document.body.innerHTML = `
        <div id="work">
            
        </div>
    `;
    const work = $('#work');
    page.initialize(document);

    const dt = DataTable.create_inside(work, {
        datasource: ArraySource.create({
            data: [
                {project: 'Generelt NT', work: '1:03:57'},
                {project: 'Tiktok',      work: '2:44:57'},
                {project: 'AFR-support', work: '1:06:43'}
            ]
        }),
        columns: {
            project: {
                label: 'Prosjekt'
            },
            work: {
                label: 'Arbeid'
            }
        }
    });
    
    console.log(dt); 
    
    // expect(dt.get_xy(1, 1).value).toEqual('2:44:57');
    expect(document.getElementById('work')).toMatchSnapshot();
    // expect(utidy(work.html())).toEqual(utidy(`
    //     <table class="DataTable dk-bx" id="data-table">
    //         <thead id="dk-bx">
    //         <tr id="dk-datatable">
    //             <th style="cursor: url(&quot;//static.datakortet.no/dk/sort-za.cur&quot;),pointer;" fieldname="project" scope="col" class="string" id="dk-datatable">
    //                 <div style="float: right; width: 2ex; margin-left: 2ex;" class="SortDirection sort-icon" id="sort-direction"></div>
    //                 Prosjekt
    //             </th>
    //             <th style="cursor: url(&quot;//static.datakortet.no/dk/sort-za.cur&quot;), pointer;" fieldname="work" scope="col" class="string" id="dk-datatable">
    //                 <div style="float: right; width: 2ex; margin-left: 2ex;" class="SortDirection sort-icon" id="sort-direction"></div>
    //                 Arbeid
    //             </th>
    //         </tr>
    //         </thead>
    //         <tfoot id="dk-datatable"></tfoot>
    //         <tbody id="dk-datatable">
    //         <tr rownum="0" pk="0" class="TableRow" id="dk-datatable">
    //             <td axis="project" class="TableCell" id="dk-tr">Generelt NT</td>
    //             <td axis="work" class="TableCell" id="dk-tr">1:03:57</td>
    //         </tr>
    //
    //         <tr rownum="1" pk="1" class="TableRow" id="dk-datatable">
    //             <td axis="project" class="TableCell" id="dk-tr">Tiktok</td>
    //             <td axis="work" class="TableCell" id="dk-tr">2:44:57</td>
    //         </tr>
    //
    //         <tr rownum="2" pk="2" class="TableRow" id="dk-datatable">
    //             <td axis="project" class="TableCell" id="dk-tr">AFR-support</td>
    //             <td axis="work" class="TableCell" id="dk-tr">1:06:43</td>
    //         </tr>
    //         </tbody>
    //     </table>
    // `));
});
