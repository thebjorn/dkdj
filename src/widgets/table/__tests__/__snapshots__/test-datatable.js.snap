// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-datatable 1`] = `
<div
  id="work"
>
  <table
    class="DataTable dk-bx"
    id="data-table-1"
  >
    <thead
      id="dk-bx-1"
    >
      

      <tr
        id="dk-table-1"
      >
        

        <th
          class="string"
          fieldname="project"
          id="dk-table-2"
          scope="col"
          style="cursor: url(\\"https://static.datakortet.no/dk/sort-az.cur\\"), pointer;"
          title=""
        >
          <div
            class="SortDirection sort-icon"
            id="sort-direction-1"
            style="float: right; width: 2ex; margin-left: 2ex;"
          >
            <i
              class="icon fa fa-chevron-down chevron-down"
              icon="chevron-down"
              value="down"
            />
          </div>
          Prosjekt
        </th>
        

        

        <th
          class="string"
          fieldname="work"
          id="dk-table-3"
          scope="col"
          style="cursor: url(\\"https://static.datakortet.no/dk/sort-za.cur\\"), pointer;"
          title=""
        >
          <div
            class="SortDirection sort-icon"
            id="sort-direction-2"
            style="float: right; width: 2ex; margin-left: 2ex;"
          />
          Arbeid
        </th>
        

      </tr>
      

    </thead>
    <tfoot
      id="dk-bx-2"
    />
    <tbody
      id="dk-bx-3"
    >
      

      <tr
        class="TableRow"
        id="dk-table-7"
        pk="1"
        rownum="0"
      >
        

        <td
          axis="project"
          class="TableCell"
          id="dk-tr-7"
        >
          Tiktok
        </td>
        

        

        <td
          axis="work"
          class="TableCell"
          id="dk-tr-8"
        >
          2:44:57
        </td>
        

      </tr>
      

      

      <tr
        class="TableRow"
        id="dk-table-8"
        pk="0"
        rownum="1"
      >
        

        <td
          axis="project"
          class="TableCell"
          id="dk-tr-9"
        >
          Generelt NT
        </td>
        

        

        <td
          axis="work"
          class="TableCell"
          id="dk-tr-10"
        >
          1:03:57
        </td>
        

      </tr>
      

      

      <tr
        class="TableRow"
        id="dk-table-9"
        pk="2"
        rownum="2"
      >
        

        <td
          axis="project"
          class="TableCell"
          id="dk-tr-11"
        >
          FOO-support
        </td>
        

        

        <td
          axis="work"
          class="TableCell"
          id="dk-tr-12"
        >
          1:06:43
        </td>
        

      </tr>
      

    </tbody>
  </table>
</div>
`;
